src/qt/trafficgraphwidget.h

```diff
@@ -1,48 +1,63 @@
-// Copyright (c) 2011-2015 The Bitcoin Core developers
-// Distributed under the MIT software license, see the accompanying
-// file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
-#ifndef BITCOIN_QT_TRAFFICGRAPHWIDGET_H
-#define BITCOIN_QT_TRAFFICGRAPHWIDGET_H
-
-#include <QWidget>
-#include <QQueue>
-
-class ClientModel;
-
-QT_BEGIN_NAMESPACE
-class QPaintEvent;
-class QTimer;
-QT_END_NAMESPACE
-
-class TrafficGraphWidget : public QWidget
-{
-    Q_OBJECT
-
-public:
-    explicit TrafficGraphWidget(QWidget *parent = 0);
-    void setClientModel(ClientModel *model);
-    int getGraphRangeMins() const;
-
-protected:
-    void paintEvent(QPaintEvent *);
-
-public Q_SLOTS:
-    void updateRates();
-    void setGraphRangeMins(int mins);
-    void clear();
-
-private:
-    void paintPath(QPainterPath &path, QQueue<float> &samples);
-
-    QTimer *timer;
-    float fMax;
-    int nMins;
-    QQueue<float> vSamplesIn;
-    QQueue<float> vSamplesOut;
-    quint64 nLastBytesIn;
-    quint64 nLastBytesOut;
-    ClientModel *clientModel;
-};
-
-#endif // BITCOIN_QT_TRAFFICGRAPHWIDGET_H
+// Copyright (c) 2011-2015 The Bitcoin Core developers
+// Copyright (c) 2022 The Dogecoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_QT_TRAFFICGRAPHWIDGET_H
+#define BITCOIN_QT_TRAFFICGRAPHWIDGET_H
+
+#include <QWidget>
+#include <QVector>
+
+class ClientModel;
+
+QT_BEGIN_NAMESPACE
+class QPaintEvent;
+class QTimer;
+QT_END_NAMESPACE
+
+class TrafficGraphWidget : public QWidget
+{
+    Q_OBJECT
+
+public:
+    explicit TrafficGraphWidget(QWidget *parent = 0);
+    void setClientModel(ClientModel *model);
+    int getGraphRangeMins() const;
+
+protected:
+    void paintEvent(QPaintEvent *);
+
+public Q_SLOTS:
+    void updateTimespanRates();
+    void updateBaseRates();
+    void setGraphRangeMins(int mins);
+    void clear();
+
+private:
+    void paintPath(QPainterPath &path, const QVector<float> &samples, int samplesHead);
+    float resampleSamples(const QVector<float> &source, QVector<float> &dest) const;
+
+    float fMax;
+    int nMins;
+

```

Comment:
This feels like the appropriate spot to add a comment about these vectors. Something like:

```
/*
 * We use these vectors as circle buffers so that we always have a maximum of BASE_SAMPLES_BUF_SIZE entries.
 * The HeadIndex variables keep track of the current position in the circle buffer.
 * When we increment these variables, we check for wraparound so we can continue filling the circle buffers,
 * continually replacing the oldest values.
 */
```

I know when I come back to read this code later, I'll appreciate the explanation here.

Reply:
Added a bit more explanation, involving how the two circular buffers interact, but also included all this.